src/chains/ethereum/ethereum/src/blockchain.ts

```diff
@@ -1121,6 +1132,446 @@ export default class Blockchain extends Emittery.Typed<
+   *  3. Create read stream to get the storage keys from the transaction
+   *  4. Sort storage keys
+   *  5. Filter, assemble, and send storage data for the given range back
+   *
+   * @param blockHash
+   * @param txIndex
+   * @param contractAddress
+   * @param startKey
+   * @param maxResult
+   */
+  public async storageRangeAt(
+    blockHash: string | Buffer,
+    txIndex: number,
+    contractAddress: string,
+    startKey: string | Buffer,
+    maxResult: number
+  ) {
+    type StorageRangeResult = {
+      nextKey: null | string;
+      storage: any;
+    };
+
+    const result: StorageRangeResult = {
+      nextKey: null,
+      storage: {}
+    };
+
+    // get block information
+    const targetBlock = await this.blocks.getByHash(blockHash);
+
+    // get parent block and use it create the state trie
+    const parentBlock = await this.blocks.getByHash(
+      targetBlock.header.parentHash.toBuffer()
+    );
+    const trie = new SecureTrie(
+      this.#database.trie,
+      parentBlock.header.stateRoot.toBuffer()
+    );
+
+    const getFromTrie = (address: Buffer): Promise<Buffer> =>
+      new Promise((resolve, reject) => {
+        trie.get(address, (err, data) => {
+          if (err) return void reject(err);
+          resolve(data);
+        });
+      });
+    const addressDataPromise = getFromTrie(
+      Address.from(contractAddress).toBuffer()
+    );
+    const addressData = await addressDataPromise;
+    if (!addressData) {
+      throw new Error(`account ${contractAddress} doesn't exist`);
+    }
+
+    // An address's stateRoot is stored in the 3rd rlp entry
+    trie.root = ((rlpDecode(addressData) as any) as [
+      Buffer /*nonce*/,
+      Buffer /*amount*/,
+      Buffer /*stateRoot*/,
+      Buffer /*codeHash*/
+    ])[2];
+
+    // use the addresses storage trie to get relevant hashed keys
+    let keys: Buffer[] = [];
+    const getStorageKeys = () => {
+      return new Promise((resolve, reject) => {
+        trie
+          .createReadStream()
+          .on("data", async data => {
+            keys.push(data.key);
+          })
+          .on("end", async () => {
+            // sort keys
+            const sortedKeys = keys.sort((a, b) => Buffer.compare(a, b));
+
+            // find starting point in array of sorted keys
+            const startKeyBuffer = Data.from(startKey).toBuffer();
+            keys = sortedKeys.filter(key => {
+              if (Buffer.compare(startKeyBuffer, key) <= 0) {
+                return key;
+              }
+            });
+
+            // only take the keys that we care about
+            keys = keys.slice(0, maxResult + 1);
+            if (keys.length > maxResult) {
+              // assign nextKey and remove it from array of keys
+              const nextKey = keys.pop();
+              result.nextKey = Data.from(nextKey).toJSON();
+            }
+
+            resolve(keys);
+          });
+      });
+    };
+    await getStorageKeys();
```

Comment:
Can you move the `let keys: Buffer[] = [];` declaration inside the `getStorageKeys` function and then:

```suggestion
    const keys = await getStorageKeys();
```